### ISSUE TITLE
Window close "X" button freezes alacritty

### ISSUE THREAD

**Issue by davidhewitt** [jwilm/alacritty#3042]
Which operating system does the issue occur on?
Windows 10 (insiders build 19033)

On my system, attempting to close Alacritty with the X button in the title bar causes it to crash. The whole application freezes. The CPU usage is 0% so I assume some kind of deadlock.

It's been a problem for a few months; not sure when it got introduced. I'd wondered about looking into it myself, but don't have the time right now. I might take a look in the New Year if it's still an issue then.

I'm running powershell as my shell. ConPTY backend.

(I'd been waiting for the glutin event loop update to land in case that fixed it. Then saw some similar work for WinPTY which I wondered if would fix it - doesn't seem to have helped.)

See example in this GIF (goes to not responding right at the end):

[![Image from Gyazo](https://i.gyazo.com/facca0b47b773434d52e3c70c547b091.gif)](https://gyazo.com/facca0b47b773434d52e3c70c547b091)

**Comment by chrisduerr** [comment]
> It's been a problem for a few months; not sure when it got introduced.

Have you been running off master, or can you confirm that this was already an issue in v0.3.3?

**Comment by davidhewitt** [comment]
Works fine in the "portable" 0.3.3. (I just downloaded to verify).

The issue is in my local compiled version. I've been running off master for some time.

**Comment by chrisduerr** [comment]
It would be very useful to try out how the winit example behaves, just to make sure this is not an issue with winit in general.

But I think it should be easy to troubleshoot this just by checking where it hangs, however I don't think I've heard any similar reports, which is a bit surprising to me.

**Comment by davidhewitt** [comment]
The "X" works correctly for the `window` example on `winit/master`.

Worth mentioning that Alacritty also closes correctly if I quit the shell with an `exit`  command instead of via the GUI. I wonder if it's related to #2695 ?

**Comment by chrisduerr** [comment]
@davidhewitt Could you try something like `sleep 10 && exit`? And then don't touch anything until it exits.

**Comment by davidhewitt** [comment]
I've just tried with the powershell command `sleep 5; exit`.

* If I don't touch _anything_, then Alacritty never closes.
* If I wait 10 seconds and then touch the mouse or keyboard, Alacritty instantly closes when I touch the the input device.
* If I wiggle my mouse the whole time, Alacritty closes after 5 seconds.

So this is an interesting one too - seems like Alacritty is waiting for one final tick of the window event loop in order to close 😄

**Comment by chrisduerr** [comment]
That's what #2695 is about, but you're saying that this is not true for this issue, right? No matter if you click or press any keys, the window doesn't close?

**Comment by davidhewitt** [comment]
Indeed. The window just hangs. Clicking on the window a few times eventually leads to the windows crash prompt:

![image](https://user-images.githubusercontent.com/1939362/69761884-99051080-1160-11ea-8f08-3267a6817b19.png)

**Comment by chrisduerr** [comment]
Then I'd assume that the issues are completely unrelated and your assumption of a deadlock somewhere is not unlikely.

We don't have many places where something like this could occur, since we basically just run two different threads, so I'd assume this wouldn't be too hard to troubleshoot with access to a Windows machine.

**Comment by ifhw** [comment]
I got same issue after  upgrade to version 0.4

**Comment by davidhewitt** [comment]
I managed to find some time to play with this on a flight. It is indeed a deadlock, in Microsoft's `ClosePseudoConsole` API.

I suspect it's an unfortunate misuse of the API by Alacritty; in the deadlocking case the in/out pipes are dropped long before `ClosePseudoConsole` is called. I suspect that some change since `0.3.3` caused the drop order to change.

I've got a WIP workaround; I'll push it in the next week or so once I've had time to finish it off and have an internet connection.
